File: new/Kotlin/home_accounting_query/src/main/kotlin/entities/FinanceRecord.kt
package com.sz.home_accounting.query.entities

import com.sz.home_accounting.query.DBException
import java.nio.ByteBuffer
import java.nio.ByteOrder

data class FinanceRecord(var operations: MutableList<FinanceOperation>) {
    companion object {
        fun fromBinary(data: ByteArray): FinanceRecord {
            val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
            var length = buffer.getShort()
            val totals = mutableMapOf<Int, Long>()
            while (length-- > 0) {
               val accountId = buffer.getShort().toInt()
               totals[accountId] = buffer.getLong()
            }
            length = buffer.getShort()
            val operations = mutableListOf<FinanceOperation>()
            while (length-- > 0) {
                operations.add(FinanceOperation.fromBinary(buffer))
            }
            if (buffer.hasRemaining()) {
                throw DBException("incorrect data")
            }
            val record = FinanceRecord(operations)
            record.totals = totals
            return record
        }

        fun create(): FinanceRecord {
            return FinanceRecord(mutableListOf())
        }
    }

    var totals = mapOf<Int, Long>()

    fun updateChanges(changes: FinanceChanges, dicts: Dicts) {
        operations.forEach { it.updateChanges(changes, dicts) }
    }

    fun buildChanges(dicts: Dicts): FinanceChanges {
        val changes = FinanceChanges(totals)
        operations.forEach {  it.updateChanges(changes, dicts) }
        return changes
    }

    fun toBinary(): ByteArray {
        val buffer = ByteBuffer.allocate(10 * 1024).order(ByteOrder.LITTLE_ENDIAN)
        buffer.putShort(totals.size.toShort())
        for ((accountId, value) in totals) {
            buffer.putShort(accountId.toShort())
            buffer.putLong(value)
        }
        buffer.putShort(operations.size.toShort())
        for (operation in operations) {
            operation.toBinary(buffer)
        }
        return buffer.array().copyOfRange(0, buffer.position())
    }
}

data class FinanceOperation(
    val amount: Long?,
    val summa: Long,
    val subcategory: Int,
    val account: Int,
    val properties: List<FinOpProperty>
) {
    companion object {
        fun fromBinary(buffer: ByteBuffer): FinanceOperation {
            val amount = buffer.getLong()
            val summa = buffer.getLong()
            val subcategory = buffer.getShort().toInt()
            val account = buffer.getShort().toInt()
            var propertiesCount = buffer.get().toInt()
            val properties = mutableListOf<FinOpProperty>()
            while (propertiesCount-- > 0) {
                properties.add(FinOpProperty.fromBinary(buffer))
            }
            return FinanceOperation(if (amount == 0L) { null } else { amount }, summa, subcategory, account, properties)
        }
    }

    fun updateChanges(changes: FinanceChanges, dicts: Dicts) {
        val subcategory = dicts.subcategories.getValue(subcategory)
        when (subcategory.operationCode) {
            SubcategoryOperationCode.Incm -> changes.income(account, summa)
            SubcategoryOperationCode.Expn -> changes.expenditure(account, summa)
            SubcategoryOperationCode.Spcl -> {
                when (subcategory.code) {
                    // Пополнение карточного счета наличными
                    SubcategoryCode.Incc -> handleIncc(changes, dicts.accounts)
                    // Снятие наличных в банкомате
                    SubcategoryCode.Expc -> handleExpc(changes, dicts.accounts)
                    // Обмен валюты
                    SubcategoryCode.Exch -> handleExch(changes)
                    // Перевод средств между платежными картами
                    SubcategoryCode.Trfr -> handleTrfr(changes)
                    else -> throw DBException("unknown subcategory code")
                }
            }
        }
    }

    private fun handleTrfr(changes: FinanceChanges) {
        handleTrfrWithSumma(changes, summa)
    }

    private fun handleExch(changes: FinanceChanges) {
        if (amount != null) {
            handleTrfrWithSumma(changes, amount / 10)
        }
    }

    private fun handleTrfrWithSumma(changes: FinanceChanges, value: Long)
    {
        if (properties.isEmpty()) return
        changes.expenditure(account, value)
        val secondAccountProperty = properties.find { it.code == FinOpPropertyCode.SECA }
        if (secondAccountProperty?.numericValue != null)
        {
            changes.income(secondAccountProperty.numericValue.toInt(), summa)
        }
    }

    private fun handleExpc(changes: FinanceChanges, accounts: Map<Int, Account>) {
        changes.expenditure(account, summa)
        // cash account for corresponding currency code
        val cashAccount = accounts.getValue(account).cashAccount
        if (cashAccount != null) {
            changes.income(cashAccount, summa)
        }
    }

    private fun handleIncc(changes: FinanceChanges, accounts: Map<Int, Account>) {
        changes.income(account, summa)
        // cash account for corresponding currency code
        val cashAccount = accounts.getValue(account).cashAccount
        if (cashAccount != null) {
            changes.expenditure(cashAccount, summa)
        }
    }

    fun toBinary(buffer: ByteBuffer) {
        buffer.putLong(amount ?: 0)
        buffer.putLong(summa)
        buffer.putShort(subcategory.toShort())
        buffer.putShort(account.toShort())
        buffer.put(properties.size.toByte())
        for (property in properties) {
            property.toBinary(buffer)
        }
    }
}

enum class FinOpPropertyCode {
    SECA,
    NETW,
    DIST,
    TYPE,
    PPTO,
    AMOU
}

data class FinOpProperty(
    val numericValue: Long?, val stringValue: String?,
    val dateValue: Int?,
    val code: FinOpPropertyCode
) {
    companion object {
        fun fromBinary(buffer: ByteBuffer): FinOpProperty {
            val code = FinOpPropertyCode.entries[buffer.get().toInt()]
            var stringValue: String? = null
            var numericValue: Long? = null
            when (code) {
                FinOpPropertyCode.SECA, FinOpPropertyCode.DIST, FinOpPropertyCode.PPTO, FinOpPropertyCode.AMOU -> {
                    numericValue = buffer.getLong()
                }
                else -> {
                    val length = buffer.get().toInt()
                    val array = ByteArray(length)
                    buffer.get(array)
                    stringValue = String(array)
                }
            }
            return FinOpProperty(numericValue, stringValue, null, code)
        }
    }

    fun toBinary(buffer: ByteBuffer) {
        buffer.put(code.ordinal.toByte())
        when (code) {
            FinOpPropertyCode.SECA, FinOpPropertyCode.DIST, FinOpPropertyCode.PPTO, FinOpPropertyCode.AMOU -> buffer.putLong(numericValue!!)
            else -> {
                val bytes = stringValue!!.toByteArray(Charsets.UTF_8)
                buffer.put(bytes.size.toByte())
                buffer.put(bytes)
            }
        }
    }
}

class FinanceChanges(totals: Map<Int, Long>) {
    val changes: MutableMap<Int, FinanceChange> =
        totals.map { it.key to FinanceChange(it.value, 0, 0) }.toMap().toMutableMap()

    fun buildTotals(): Map<Int, Long> {
        return changes.map { it.key to it.value.getEndSumma() }.toMap()
    }

    fun income(account: Int, value: Long) {
        changes.getOrPut(account) { FinanceChange(0, 0, 0) }.income += value
    }

    fun expenditure(account: Int, value: Long) {
        changes.getOrPut(account) { FinanceChange(0, 0, 0) }.expenditure += value
    }
}

data class FinanceChange(val summa: Long, var income: Long, var expenditure: Long) {
    fun getEndSumma(): Long {
        return summa + income - expenditure
    }
}

File: new/dotNET/HomeAccountingServiceClient/HomeAccountingServiceClientLibrary/entities/FinanceRecord.cs
using System.Text;

namespace HomeAccountingServiceClientLibrary.entities;

public enum FinOpPropertyCode {
    Seca,
    Netw,
    Dist,
    Type,
    Ppto,
    Amou
}

public class FinanceRecord
{
    public Dictionary<int, long> Totals;
    public List<FinanceOperation> Operations { get; }

    public FinanceRecord(List<FinanceOperation> operations, Dictionary<int, long> totals)
    {
        Totals = totals;
        Operations = operations;
    }
    
    public FinanceRecord(): this([], new Dictionary<int, long>())
    {
    }

    internal static FinanceRecord FromBinary(byte[] data)
    {
        using var reader = new BinaryReader(new MemoryStream(data));
        
        var length = reader.ReadInt16();
        var totals = new Dictionary<int, long>();
        while (length-- > 0)
        {
            var accountId = (int)reader.ReadInt16();
            totals[accountId] = reader.ReadInt64();
        }

        length = reader.ReadInt16();
        var operations = new List<FinanceOperation>();
        while (length-- > 0)
            operations.Add(FinanceOperation.FromBinary(reader));
        if (reader.BaseStream.Position != reader.BaseStream.Length)
            throw new DbException("incorrect data");
        return new FinanceRecord(operations, totals);
    }
    
    internal void UpdateChanges(FinanceChanges changes, Dicts dicts)
    {
        Operations.ForEach(op => op.UpdateChanges(changes, dicts));
    }

    internal FinanceChanges BuildChanges(Dicts dicts)
    {
        var changes = new FinanceChanges(Totals);
        Operations.ForEach(op => op.UpdateChanges(changes, dicts));
        return changes;
    }
    
    internal FinanceRecord BuildNextRecord(Dicts dicts)
    {
        return new FinanceRecord([], BuildChanges(dicts).BuildTotals());
    }
}

public record FinanceOperation(long? Amount, long Summa, int SubcategoryId, int AccountId, List<FinOpProperty> Properties)
{
    internal static FinanceOperation FromBinary(BinaryReader reader)
    {
        var amount = reader.ReadInt64();
        var summa = reader.ReadInt64();
        var subcategoryId = (int)reader.ReadInt16();
        var accountId = (int)reader.ReadInt16();
        var propertiesCount = (int)reader.ReadByte();
        var properties = new List<FinOpProperty>();
        while (propertiesCount-- > 0)
            properties.Add(FinOpProperty.FromBinary(reader));
        return new FinanceOperation(amount == 0 ? null : amount, summa, subcategoryId, accountId, properties);
    }
    
   internal void UpdateChanges(FinanceChanges changes, Dicts dicts)
    {
        var subcategory = dicts.Subcategories[SubcategoryId];
        switch (subcategory.OperationCode)
        {
            case SubcategoryOperationCode.Incm:
                changes.Income(AccountId, Summa);
                break;
            case SubcategoryOperationCode.Expn:
                changes.Expenditure(AccountId, Summa);
                break;
            case SubcategoryOperationCode.Spcl:
                switch (subcategory.Code)
                {
                    // Пополнение карточного счета наличными
                    case SubcategoryCode.Incc:
                        HandleIncc(changes, dicts.Accounts);
                        break;
                    // Снятие наличных в банкомате
                    case SubcategoryCode.Expc:
                        HandleExpc(changes, dicts.Accounts);
                        break;
                    // Обмен валюты
                    case SubcategoryCode.Exch:
                        HandleExch(changes);
                        break;
                    // Перевод средств между платежными картами
                    case SubcategoryCode.Trfr:
                        HandleTrfr(changes);
                        break;
                    default: throw new DbException("unknown subcategory code");
                }
                break;
            default: throw new DbException("unknown subcategory operation code");
        }
    }

    private void HandleTrfr(FinanceChanges changes)
    {
        HandleTrfrWithSumma(changes, Summa);
    }

    private void HandleExch(FinanceChanges changes)
    {
        if (Amount == null)
            throw new DbException("null summa");
        HandleTrfrWithSumma(changes, Amount.Value / 10);
    }

    private void HandleTrfrWithSumma(FinanceChanges changes, long summa)
    {
        var secondAccountProperty = Properties.Find(property => property.Code == FinOpPropertyCode.Seca);
        var seca = secondAccountProperty?.NumericValue;
        if (seca == null)
            throw new DbException("null ssecond account");
        changes.Expenditure(AccountId, summa);
        changes.Income((int)seca, Summa);
    }
    
    private void HandleExpc(FinanceChanges changes, Dictionary<int, Account> accounts)
    {
        changes.Expenditure(AccountId, Summa);
        // cash account for corresponding currency code
        var cashAccount = accounts[AccountId].CashAccount;
        if (cashAccount == null)
            throw new DbException("null cash account");
        changes.Income((int)cashAccount, Summa);
    }

    private void HandleIncc(FinanceChanges changes, Dictionary<int, Account> accounts)
    {
        changes.Income(AccountId, Summa);
        // cash account for corresponding currency code
        var cashAccount = accounts[AccountId].CashAccount;
        if (cashAccount == null)
            throw new DbException("null cash account");
        changes.Expenditure((int)cashAccount, Summa);
    }}

public record FinOpProperty(long? NumericValue, string? StringValue, int? DateValue, FinOpPropertyCode Code)
{
    internal static FinOpProperty FromBinary(BinaryReader reader)
    {
        var code = (FinOpPropertyCode)reader.ReadByte();
        string? stringValue = null;
        long? numericValue = null;
        switch (code)
        {
            case FinOpPropertyCode.Seca:
            case FinOpPropertyCode.Dist:
            case FinOpPropertyCode.Ppto:
            case FinOpPropertyCode.Amou:
                numericValue = reader.ReadInt64();
                break;
            default:
                var length = (int)reader.ReadByte();
                stringValue = Encoding.ASCII.GetString(reader.ReadBytes(length));
                break;
        }

        return new FinOpProperty(numericValue, stringValue, null, code);
    }
}

internal class FinanceChanges
{
    internal readonly Dictionary<int, FinanceChange> Changes;

    internal FinanceChanges(Dictionary<int, long> totals)
    {
        Changes = totals
            .Select(kv => (kv.Key, new FinanceChange(kv.Value)))
            .ToDictionary();
    }

    internal void Income(int account, long summa)
    {
        if (Changes.TryGetValue(account, out var change))
            change.Income += summa;
        else 
            Changes[account] = new FinanceChange(0, summa, 0);
    }
    
    internal void Expenditure(int account, long summa)
    {
        if (Changes.TryGetValue(account, out var change))
            change.Expenditure += summa;
        else 
            Changes[account] = new FinanceChange(0, 0, summa);
    }

    public Dictionary<int, long> BuildTotals()
    {
        return Changes
            .Select(kv => (kv.Key, kv.Value.GetEndSumma()))
            .ToDictionary();
    }

    public void Cleanup(HashSet<int> accountIds)
    {
        var keysToDelete = Changes.Keys.Except(accountIds);
        foreach (var key in keysToDelete)
            Changes.Remove(key);
    }
}

internal class FinanceChange
{
    internal readonly long Summa;
    internal long Income;
    internal long Expenditure;

    internal FinanceChange(long summa)
    {
        Summa = summa;
        Income = Expenditure = 0;
    }

    internal FinanceChange(long summa, long income, long expenditure)
    {
        Summa = summa;
        Income = income;
        Expenditure = expenditure;
    }

    internal long GetEndSumma()
    {
        return Summa + Income - Expenditure;
    }
}
